plutus-emulator/src/Wallet/Emulator/Wallet.hs

```diff
@@ -0,0 +1,235 @@
+    UpdatePaymentWithChange :: Value -> (Set.Set TxIn, Maybe TxOut) -> WalletEffect (Set.Set TxIn, Maybe TxOut)
+    WatchedAddresses :: WalletEffect AM.AddressMap
+    StartWatching :: Address -> WalletEffect ()
+    WalletSlot :: WalletEffect Slot
+    WalletNotify :: Notification -> WalletEffect ()
+    WalletLogMsg :: T.Text -> WalletEffect ()
+makeEffect ''WalletEffect
+
+type WalletEffs = '[NC.NodeClientEffect, State WalletState, Error WAPI.WalletAPIError, Writer [WalletEvent]]
+
+handleWallet
+    :: (Members WalletEffs effs)
+    => Eff (WalletEffect ': effs) ~> Eff effs
+handleWallet = interpret $ \case
+    SubmitTxn tx -> NC.publishTx tx
+    OwnPubKey -> toPublicKey <$> gets _ownPrivateKey
+    Sign bs -> do
+        privK <- gets _ownPrivateKey
+        pure (Crypto.sign (BSL.toStrict bs) privK)
+    UpdatePaymentWithChange vl (oldIns, changeOut) -> do
+        ws <- get
+        let -- These inputs have been already used, we won't touch them
+            usedFnds = Set.map txInRef oldIns
+            -- Optional, left over change. Replace a `Nothing` with a Value of 0.
+            oldChange = maybe (Ada.lovelaceValueOf 0) txOutValue changeOut
+            -- Available funds.
+            fnds   = Map.withoutKeys (ws ^. ownFunds) usedFnds
+            privK  = ws ^. ownPrivateKey
+            pubK   = toPublicKey privK
+        if vl `Value.leq` oldChange
+        then
+          -- If the requested value is covered by the change we only need to update
+          -- the remaining change.
+          pure (oldIns, mkChangeOutput pubK $ oldChange PlutusTx.- vl)
+        else do
+          -- If the requested value is not covered by the change, then we need to
+          -- select new inputs, after deducting the oldChange from the value.
+          (spend, change) <- selectCoin (second (txOutValue . txOutTxOut) <$> Map.toList fnds)
+                                        (vl PlutusTx.- oldChange)
+          let ins = Set.fromList (pubKeyTxIn pubK . fst <$> spend)
+          pure (Set.union oldIns ins, mkChangeOutput pubK change)
+    WatchedAddresses -> gets _addressMap
+    StartWatching a -> modify $ \ws -> ws & addressMap %~ AM.addAddress a
+    WalletSlot -> NC.currentSlot
+    WalletNotify n -> case n of
+        CurrentSlot _       -> pure ()
+        BlockValidated blck -> mapM_ (\t -> modify $ \ws -> ws & addressMap %~ AM.updateAllAddresses t) blck
+    WalletLogMsg m -> tell [WalletMsg m]
+
+-- HACK: these shouldn't exist, but WalletAPI needs to die first
+instance (Member WalletEffect effs) => WAPI.WalletAPI (Eff effs) where
+    submitTxn = submitTxn
+    ownPubKey = ownPubKey
+    sign = sign
+    updatePaymentWithChange = updatePaymentWithChange
+    watchedAddresses = watchedAddresses
+    startWatching = startWatching
+    slot = walletSlot
+
+instance (Member (Error WAPI.WalletAPIError) effs) => E.MonadError WAPI.WalletAPIError (Eff effs) where
+    throwError = throwError
+    catchError = catchError
```

Comment:
It's not a loop, because lhses come from `Control.Monad.Except` and rhses from `freer-simple`? Worth a short note, I think, because it looks like a loop.

Reply:
Yes, names inside typeclass instances are a bit funny.